### PR TITLE
Enabling an instance of a Tree class to a second container - Making scalar types immutable

### PR DESCRIPTION
**Pull request by Armanbqt**
*Issue #52*

*Description of changes:*
- Removed IonValue.Container
- Removed setters for scalar types in IonValue and its derived classes
- Removed ContainedValueException

### CODE REVIEW

File: IonDotnet.Tests/Tree/IonSequenceTest.cs [hunk 1]
```diff
@@ -48,7 +48,7 @@ public void Insert(int idx, int count)
             }
 
             v.Insert(idx, r);
-            Assert.AreEqual(v, r.Container);
+            //Assert.AreEqual(v, r.Container);
```
Comment (tgregg) [inline IonDotnet.Tests/Tree/IonSequenceTest.cs]: Remove commented-out line.

File: IonDotnet.Tests/Tree/IonSequenceTest.cs [hunk 2]
```diff
@@ -71,7 +71,7 @@ public void RemoveAt(int idx, int count)
             var r = v[idx];
             v.RemoveAt(idx);
             Assert.AreEqual(count - 1, v.Count);
-            Assert.IsNull(r.Container);
+            //Assert.IsNull(r.Container);
```
Comment (tgregg) [inline IonDotnet.Tests/Tree/IonSequenceTest.cs]: Remove commented-out line.

File: IonDotnet/Internals/IonTreeWriter.cs [hunk 3]
```diff
@@ -9,11 +11,15 @@ namespace IonDotnet.Internals
     internal class IonTreeWriter : IonSystemWriter
     {
         private IIonContainer _currentContainer;
+        private Stack<IIonContainer> _containers = new Stack<IIonContainer>();
+        // Holds pairs of  containers, and their respective parents
+        private IDictionary _containersParents = new Dictionary<IIonValue, IIonContainer>();
```
Comment (tgregg) [inline IonDotnet/Internals/IonTreeWriter.cs]: Is this needed?

File: IonDotnet/Tree/Impl/IonSequence.cs [hunk 4]
```diff
@@ -87,12 +84,9 @@ public override bool Remove(IIonValue item)
         {
             ThrowIfLocked();
             ThrowIfNull();
-            if (NullFlagOn() || item?.Container != this)
```
Comment (tgregg) [inline IonDotnet/Tree/Impl/IonSequence.cs]: Do we still need the `NullFlagOn()` check?

File: IonDotnet/Internals/IonTreeWriter.cs (hunk 3 above)
Comment (Armanbqt) [reply IonDotnet/Internals/IonTreeWriter.cs]: You're right, we can just use the stack

File: IonDotnet/Tree/Impl/IonSequence.cs (hunk 4 above)
Comment (Armanbqt) [reply IonDotnet/Tree/Impl/IonSequence.cs]: ```NullFlagOn()``` is in IonValue and is used by different classes for different checks. For example ```IsEquivalentTo```

Comment (tgregg) [reply IonDotnet/Tree/Impl/IonSequence.cs]: Sure. What I mean is that the removal of the `NullFlagOn()` check at this location doesn't seem related to the rest of the CR. Was that check never needed here, or is it related in some way that I don't understand?

Comment (Armanbqt) [reply IonDotnet/Tree/Impl/IonSequence.cs]: I see, sorry I understood it wrong.
Seems like I shouldn't have removed it.

**Pull request merged**